hermes/probe/checknil/check_nil.go

```diff
@@ -0,0 +1,332 @@
+			return probe.NewProbeError(metrics.Success, nil)
+		}
+	})
+
+	logger.Infof("checkNilFileExists(): nil file located on target storage system.")
+	return file, err
+}
+
+// checkNilFileExists checks that the nil file exists on the target system.
+// Arguments:
+//	- ctx: context so this probe can be cancelled if needed.
+//	- target: target run information.
+//	- client: initialised storage client for this target system.
+//	- logger: logger associated with the probe calling this function.
+// Returns:
+//	- nilFile: a nilFile struct with the file name recorded
+//	- err: error exit statuses are well defined in metrics.go
+func readNilFile(ctx context.Context, file *nilFile, bucket string, target *probe.Target, client stiface.Client, logger *logger.Logger) (*nilFile, *probe.ProbeError) {
+	fileObject := client.Bucket(bucket).Object(file.name)
+	var reader stiface.Reader
+
+	err := RecordAPILatency(target.LatencyMetrics.APICallLatency, metrics.APIGetFile, func() *probe.ProbeError {
+		var err error
+		reader, err = fileObject.NewReader(ctx)
+		if err != nil {
+			return probe.NewProbeError(metrics.FileReadFailure, fmt.Errorf("readNilFile(): unable to read nil file: %w", err))
+
+		}
+		return probe.NewProbeError(metrics.Success, nil)
+	})
+	defer reader.Close()
+
+	if err.Status != metrics.Success {
+		return nil, err
+	}
+
+	logger.Infof("readNilFile(): nil file retrieved from target storage system.")
+
+	var readErr error
+	file.contents, readErr = ioutil.ReadAll(reader)
+	if readErr != nil {
+		return nil, probe.NewProbeError(metrics.FileReadFailure, fmt.Errorf("readNilFile(): unable to read nil file: %w", readErr))
+	}
+
+	logger.Infof("readNilFile(): nil file read complete.")
+
+	return file, probe.NewProbeError(metrics.Success, nil)
+}
+
+// verifyNilFile verifies that the contents of the nil file match its checksum
+// and that the journal matches the existing in-memory proto.
+// Arguments:
+//	- ctx: context so this probe can be cancelled if needed.
+//	- nilFile: a nilFile struct with the file name and contents recorded
+//	- target: target run information.
+//	- client: initialised storage client for this target system.
+//	- logger: logger associated with the probe calling this function.
+// Returns:
+//	- err: error exit statuses are well defined in metrics.go
+func verifyNilFile(ctx context.Context, file *nilFile, target *probe.Target, client stiface.Client, logger *logger.Logger) *probe.ProbeError {
+	hash := sha1.New()
+	if _, err := io.Copy(hash, file.reader()); err != nil {
+		return probe.NewProbeError(metrics.ProbeFailed, fmt.Errorf("verifyNilFile(): unable to compute checksum of file contents: %w", err))
+	}
+	file.hash = fmt.Sprintf("%x", hash.Sum(nil))
+
+	if !file.validateChecksum() {
+		return probe.NewProbeError(metrics.FileCorrupted, fmt.Errorf("verifyNilFile(): checksum of nil file did not match filename"))
+	}
+
+	journal := &sjpb.StateJournal{}
+	if err := proto.Unmarshal(file.contents, journal); err != nil {
```

Comment:
`verifyNilFile()` is doing a lot of unrelated work.
I would split it up to these phases:

1. Retrieve the file
1.1 Get the file
1.2 Unmarshal the journal
2. Validate
2.1 Validate the checksum
2.2 Validate the journal

Reply:
I have refactored this into 3 functions: one that modifies the contents of the journalFile. and two that perform operations on the journalFile but do not modify its contents.